text/2603-rust-symbol-name-mangling-v0.md

```diff
@@ -724,16 +726,40 @@ Mangled names conform to the following grammar:
 <dyn-bounds> = [<binder>] {<dyn-trait>} "E"
 <dyn-trait> = <path> {<dyn-trait-assoc-binding>}
 <dyn-trait-assoc-binding> = "p" <undisambiguated-identifier> <type>
-<const> = <type> <const-data>
-        | "p" // placeholder, shown as _
+
+// Constants are encoded structurally, as a tree of array/tuple/ADT constructors,
+// with integer(-like) leaves, not using the constant's memory representation.
+// See the comments on <const-int> & <const-str> for more details on leaf encoding.
+<const> = <int-type> <const-int>
+        | "b" <const-int>           // false, true
+        | "c" <const-int>           // '...'
+        | "e" <const-str>           // "..."
+        | "R" <const>               // &value
+        | "Q" <const>               // &mut value
+        | "A" {<const>} "E"         // [a, b, c, ...]
+        | "T" {<const>} "E"         // (a, b, c, ...)
+        | "V" <path> <const-fields> // named struct/variant
+        | "p"                       // placeholder, shown as _
         | <backref>
 
-// The encoding of a constant depends on its type. Integers use their value,
-// in base 16 (0-9a-f), not their memory representation. Negative integer
-// values are preceded with "n". The bool value false is encoded as `0_`, true
-// value as `1_`. The char constants are encoded using their Unicode scalar
-// value.
-<const-data> = ["n"] {<hex-digit>} "_"
+<const-fields> = "U"                            // X
+               | "T" {<const>} "E"              // X(a, b, c, ...)
+               | "S" {<identifier> <const>} "E" // X { field: value, ... }
```

Comment:
I'm not sure why `U` is necessary, `UnitStruct` and `UnitStruct {}` are the same constant, so `U` can be `S {} E`, the only benefit is slight compression.

I'm also not sure why identifiers for field names are necessary, fields are uniquely identifiable by their indices, so `"S" {<const>} "E"` should work equally well?

In other words, `"S" {<const>} "E"` appears to be usable for any structs and variants.

Reply:
We could also "just" flatten the constant and only put the leaf values in, we wouldn't need any names or tuple/array distinctions and nesting.

The only reason we tell apart any of these things is being able to recover useful syntax for the users, and I somewhat assumed that to be a given, I suppose.

I'm obviously biased towards the approach I implemented, but maybe you could register a concern with rfcbot, that could get discussed by compiler team? (though I'm not sure in what kind of meeting)

Reply:
Done.

Reply:
I would say we keep the `U` just in order to make it consistent: we already encode `()` as `u` where we could also encode it as an empty tuple `TE`.